JASP-Engine/JASP/R/mediationanalysis.R

```diff
@@ -0,0 +1,573 @@
+    restab$addColumnInfo(name = "se",       title = "Std. Error", type = "number", format = "sf:4;dp:3")
+    restab$addColumnInfo(name = "z",        title = "z-value",    type = "number", format = "sf:4;dp:3")
+    restab$addColumnInfo(name = "pvalue",   title = "p",          type = "number", format = "dp:3;p:.001")
+    restab$addColumnInfo(name = "ci.lower", title = "Lower",      type = "number", format = "sf:4;dp:3",
+                         overtitle = paste0(options$ciWidth * 100, "% Confidence Interval"))
+    restab$addColumnInfo(name = "ci.upper", title = "Upper",      type = "number", format = "sf:4;dp:3",
+                         overtitle = paste0(options$ciWidth * 100, "% Confidence Interval"))
+    restab$addFootnote(se_message)
+    pe_res <- pe[pe$op == "~~" &
+                   pe$lhs != pe$rhs &
+                   !.unv(pe$lhs) %in% options$predictor &
+                   !.unv(pe$lhs) %in% options$confounds,]
+
+    restab[["lhs"]]      <- .unv(pe_res$lhs)
+    restab[["op"]]       <- rep("\u2194", nrow(pe_res))
+    restab[["rhs"]]      <- .unv(pe_res$rhs)
+    restab[["est"]]      <- pe_res$est
+    restab[["se"]]       <- pe_res$se
+    restab[["z"]]        <- pe_res$z
+    restab[["pvalue"]]   <- pe_res$pvalue
+    restab[["ci.lower"]] <- pe_res$ci.lower
+    restab[["ci.upper"]] <- pe_res$ci.upper
+  }
+
+}
+
+.medRsquared <- function(jaspResults, medResult, options, errors) {
+  if (!options$rsquared || !is.null(jaspResults[["rsquared"]])) return()
+  
+  jaspResults[["rsquared"]] <- tabr2 <- createJaspTable("R-Squared")
+  tabr2$addColumnInfo(name = "__var__", title = "", type = "string")
+  tabr2$addColumnInfo(name = "rsq", title = "R\u00B2", type = "number", format = "sf:4;dp:3")
+  tabr2$setExpectedSize(rows = 1, cols = 2)
+  tabr2$dependOn(options = "rsquared", optionsFromObject = jaspResults[["stateMedResult"]])
+  tabr2$position <- 1
+  if (is.null(medResult)) return()
+  
+  r2res              <- lavaan::inspect(medResult, "r2")
+  tabr2[["__var__"]] <- .unv(names(r2res))
+  tabr2[["rsq"]]     <- r2res
+}
+
+.medPathPlot <- function(jaspResults, medResult, options, errors) {
+  if (!options$pathplot || !is.null(errors)) return()
+
+  # compute the layout from the options
+  n_pred <- length(options$predictor)
+  n_medi <- length(options$mediators)
+  n_deps <- length(options$dependent)
+  n_conf <- length(options$confounds)
+
+  n_totl <- n_pred + n_medi + n_deps + n_conf
+  pred_l <- cbind(rep(-1, n_pred), .medPathLayout(n_pred))
+  deps_l <- cbind(rep(1,  n_deps), .medPathLayout(n_deps))
+  if (n_medi == 1 && n_pred == 1 && n_deps == 1) {
+    medi_l <- cbind(0, .25)
+  } else {
+    medi_l <- cbind(rep(0,  n_medi), .medPathLayout(n_medi))
+  }
+  if (n_conf > 0) {
+    conf_l <- cbind(rep(-2, n_conf), .medPathLayout(n_conf) + 0.25)
+  } else {
+    conf_l <- NULL
+  }
+  # creat a qgraph object using semplot
+  png()
+  pp <- semPlot::semPaths(
```

Comment:
I added a convenience function for this that removes the need for `png() -> dev.off()` (which has the added benefit that crashes in the code don't leave a device open)

```
pp <- .suppressGrDevice(semPlot::semPaths(
    object         = .medLavToPlotObj(medResult),
    layout         = rbind(deps_l, medi_l, pred_l, conf_l),
    intercepts     = FALSE,
    reorder        = FALSE,
    whatLabels     = ifelse(options$plotpars, "par", "name"),
    edge.color     = "black",
    color          = list(lat = "#EAEAEA", man = "#EAEAEA", int = "#FFFFFF"),
    border.width   = 1.5,
    edge.label.cex = 0.9,
    lty            = 2,
    title          = FALSE,
    sizeMan        = round(8*exp(-n_totl/80)+1)
  ))
```